core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessageDuplicator.java

```diff
@@ -657,4 +750,139 @@ public String toString() {
+            final int removalRequestedOffset = lastRemovalRequestedOffset;
+            final int numElementsToBeRemoved = removalRequestedOffset - headOffset;
+            final int bitMask = elements.length - 1;
+            final int oldHead = head;
+            int removedLength = 0;
+            for (int numRemovals = 0; numRemovals < numElementsToBeRemoved; numRemovals++) {
+                final int index = (oldHead + numRemovals) & bitMask;
+                final Object o = elements[index];
+                if (!(o instanceof CloseEvent)) {
+                    removedLength += signalLengthGetter.length(o);
+                }
+                ReferenceCountUtil.safeRelease(o);
+                elements[index] = null;
+            }
+            head = (oldHead + numElementsToBeRemoved) & bitMask;
+            headOffset = removalRequestedOffset;
+            size = size - numElementsToBeRemoved;
+            return removedLength;
+        }
+
+        private void doubleCapacity() {
+            assert head == tail;
+            final int h = head;
+            final Object[] elements = this.elements;
+            final int n = elements.length;
+            final int r = n - h; // number of elements to the right of h including h
+            final int newCapacity = n << 1;
+            if (newCapacity < 0) {
+                throw new IllegalStateException("published more than Integer.MAX_VALUE signals.");
+            }
+            final Object[] a = new Object[newCapacity];
+            final int hOffset = headOffset;
+            if ((hOffset & newCapacity - 1) == (hOffset & n - 1)) { // even number head wrap-around
+                // [4, 5, 6, 3] will be [N, N, N, 3, 4, 5, 6, N]
+                System.arraycopy(elements, h, a, h, r); // copy 3
+                System.arraycopy(elements, 0, a, n, h); // copy 4,5,6
+                tail = tail + n;
+            } else { // odd number head wrap-around
+                // [8, 5, 6, 7] will be [8, N, N, N, N, 5, 6, 7]
+                System.arraycopy(elements, h, a, h + n, r); // copy 5,6,7
+                System.arraycopy(elements, 0, a, 0, h); // copy 8
+                head = h + n;
+            }
+            this.elements = a;
+        }
+
+        Object get(int offset) {
+            final int head = this.head;
+            final int tail = this.tail;
+            final int length = elements.length;
+            final int convertedIndex = offset & (length - 1);
+            checkState(size > 0, "queue is empty");
+            checkArgument(head < tail ? head <= convertedIndex && convertedIndex < tail
+                                      : (head <= convertedIndex && convertedIndex < length) ||
+                                        (0 <= convertedIndex && convertedIndex < tail),
+                          "index: %s is invalid. head: %s, tail: %s, capacity: %s ",
+                          offset, head, tail, length);
+            return elements[convertedIndex];
```

Comment:
Well, I have thought that it doesn't need.. But I now, think it's better to have it.